### PR TITLE
imlib/draw: Add anti aliasing in draw_ellipse.

### PR DESCRIPTION
**Pull request by xue-singtown**
Hi, I add anti aliasing feature in draw_line method. I tested Binary, Grayscale and RGB565 format.

Reference code: https://gist.github.com/randvoorhies/807ce6e20840ab5314eb7c547899de68

test code in OpenMV Cam:

```
import sensor
sensor.reset() 
sensor.set_pixformat(sensor.RGB565)
sensor.set_framesize(sensor.QVGA)
sensor.skip_frames(time=2000)

a = 11
b = 60
angle = 30

while True:
    img = sensor.snapshot()

    img.draw_ellipse(50, 80, a, b, angle, fill=False, color=(255, 0, 0), thickness=1)
    img.draw_ellipse(100, 80, a, b, angle, fill=False, color=(255, 0, 0), thickness=5)
    img.draw_ellipse(150, 80, a, b, angle, fill=True, color=(255, 0, 0), thickness=5)

```
img_degree_30

# Known issue

There are drawing errors in draw ellipse with width:
1. There will be extra pixels at the corners.
2. It will become a straight line at a certain angle.

img_degree_17

Note: the function of drawing an ellipse with a line width of 1 and the function of drawing a line width of other than 1 are two functions. Thickness = 1 is perfect.

I've been researching for a few days and can't figure this out. If you guys can figure it out, I can help.

### CODE REVIEW

**Comment by iabdalkader** [comment]
Please fix the comments, they should start with a capital letter, end with a ".".
```diff
-// plot ellipse rotated by angle (degree)  👎🏼 
+// Plot ellipse rotated by angle (degree). 👍🏼 
```

File: src/omv/imlib/draw.c [hunk 1]
```diff
@@ -328,6 +327,285 @@ void imlib_draw_circle(image_t *img, int cx, int cy, int r, int c, int thickness
...
+        do {
+            // Pixel loop.
+            cur = IM_MIN(dx - xy, xy - dy);
+            ed = IM_MAX(dx - xy, xy - dy);
+            ed += 2 * ed * cur * cur / (4. * ed * ed + cur * cur); // Approximate error distance.
+            x1 = 256 * fast_fabsf(err - dx - dy + xy) / ed; // Get blend value by pixel error.
+            if (x1 < 256) {
+                imlib_set_pixel_aa(img, x0, y0, x1, c);                          // Plot curve.
+            }
+            f = (2 * err + dy < 0);
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:380]:
```suggestion
            bool f = (2 * err + dy < 0);
```

File: src/omv/imlib/draw.c [hunk 2]
```diff
@@ -328,6 +327,285 @@ void imlib_draw_circle(image_t *img, int cx, int cy, int r, int c, int thickness
...
+
+        if (w < 0.5 && (dx + xx <= 0 || dy + yy >= 0)) {
+            // Flat ellipse, algo fails.
+            cur = (w + 1.0) / 2.0; w = fast_sqrtf(w); xy = 1.0 / (w + 1.0);
+            sx = fast_floorf((x0 + 2.0 * w * x1 + x2) * xy / 2.0 + 0.5); // Subdivide curve.
+            sy = fast_floorf((y0 + 2.0 * w * y1 + y2) * xy / 2.0 + 0.5); // Plot separately.
+            dx = fast_floorf((w * x1 + x0) * xy + 0.5); dy = fast_floorf((y1 * w + y0) * xy + 0.5);
+            plot_quad_rational_bezier_width_seg(img, x0, y0, dx, dy, sx, sy, cur, c, th);
+            dx = fast_floorf((w * x1 + x2) * xy + 0.5); dy = fast_floorf((y1 * w + y2) * xy + 0.5);
+            return plot_quad_rational_bezier_width_seg(img, sx, sy, dx, dy, x2, y2, cur, c, th);
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:487]:
```suggestion
            cur = (w + 1.0) / 2.0;
            w = fast_sqrtf(w);
            xy = 1.0 / (w + 1.0);
            sx = fast_floorf((x0 + 2.0 * w * x1 + x2) * xy / 2.0 + 0.5); // Subdivide curve.
            sy = fast_floorf((y0 + 2.0 * w * y1 + y2) * xy / 2.0 + 0.5); // Plot separately.
            dx = fast_floorf((w * x1 + x0) * xy + 0.5);
            dy = fast_floorf((y1 * w + y0) * xy + 0.5);
            plot_quad_rational_bezier_width_seg(img, x0, y0, dx, dy, sx, sy, cur, c, th);
            dx = fast_floorf((w * x1 + x2) * xy + 0.5);
            dy = fast_floorf((y1 * w + y2) * xy + 0.5);
            return plot_quad_rational_bezier_width_seg(img, sx, sy, dx, dy, x2, y2, cur, c, th);
```

File: src/omv/imlib/draw.c [hunk 3]
```diff
@@ -328,6 +327,285 @@ void imlib_draw_circle(image_t *img, int cx, int cy, int r, int c, int thickness
...
+    if (w != 0) {
+        w = (w - zd) / (w + w);          // Squared weight of p1.
+    }
+    if (!(w <= 1.0 && w >= 0.0)) {
+        return;               // Limit angle to |zd|<=xd*yd.
+    }
+
+    xd = fast_floorf(xd * w + 0.5);
+    yd = fast_floorf(yd * w + 0.5);  // Snap to int.
+
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:586]:
```suggestion
```

File: src/omv/imlib/draw.c [hunk 4]
```diff
@@ -328,6 +327,285 @@ void imlib_draw_circle(image_t *img, int cx, int cy, int r, int c, int thickness
...
+                    y1 = y0 - y1 * sy;                         // Start point.
+                    imlib_set_pixel_aa(img, x0, y1, 256 * e2 / ed, c);  // Aliasing pre-pixel.
+                    for (e2 = dx - e2 - w; e2 + dx < ed; e2 += dx) {
+                        imlib_set_pixel(img, x0, y1 += sy, c);        // Pixel on thick line.
+                    }
+                    imlib_set_pixel_aa(img, x0, y1 + sy, 256 * e2 / ed, c); // Aliasing post-pixel.
+                    if (x0 == x2) {
+                        return;               // Last pixel -> curve finished.
+                    }
+                    x0 += sx; dx += xy; err += dy; dy += yy;        // X step.
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:546]:
```suggestion
                    // X step.
                    x0 += sx;
                    dx += xy;
                    err += dy;
                    dy += yy;
```

File: src/omv/imlib/draw.c [hunk 5]
```diff
@@ -328,6 +327,285 @@ void imlib_draw_circle(image_t *img, int cx, int cy, int r, int c, int thickness
...
+        }
+        xx = 2.0 * (4.0 * w * sx * xx + dx * dx);   // Differences 2nd degree.
+        yy = 2.0 * (4.0 * w * sy * yy + dy * dy);
+        sx = x0 < x2 ? 1 : -1;                            // X step direction.
+        sy = y0 < y2 ? 1 : -1;                            // Y step direction.
+        xy = -2.0 * sx * sy * (2.0 * w * xy + dx * dy);
+
+        if (cur * sx * sy < 0) {
+            // Negated curvature.
+            xx = -xx; yy = -yy; cur = -cur; xy = -xy;
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:355]:
```suggestion
            xx = -xx;
            yy = -yy;
            cur = -cur;
            xy = -xy;
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]
Please avoid multiple variable definitions, function calls, and assignments on a single line. I fixed a few examples, but there are more in code.

**Comment by kwagyeman** [comment]
@xue-singtown @iabdalkader - I'll see if I can process this PR this week.

**Comment by kwagyeman** [comment]
@xue-singtown - I am working on fixing this PR now.

**Comment by kwagyeman** [comment]
Replaced by https://github.com/openmv/openmv/pull/2195.

**Review by kwagyeman — COMMENTED** [review]
Replaced by: https://github.com/openmv/openmv/pull/2195

I managed to add fill support natively to the code. I don't quite see how to fix the weird graphical glitches it has when changing from steep to flat and vice versus curves.

**Pull request closed without merging**